File: server/src/routes/feedback.js
const express = require('express');
const router = express.Router();
const fs = require('fs');
const path = require('path');

const FEEDBACK_FILE = path.join(__dirname, '../../feedback_log.json');

// Helper to read feedback
function getFeedbackList() {
  try {
    if (fs.existsSync(FEEDBACK_FILE)) {
      const data = fs.readFileSync(FEEDBACK_FILE, 'utf8');
      return JSON.parse(data || '[]');
    }
  } catch (e) {
    console.error('Error reading feedback_log.json:', e);
  }
  return [];
}

// Helper to save feedback
function saveFeedbackList(list) {
  try {
    fs.writeFileSync(FEEDBACK_FILE, JSON.stringify(list, null, 2), 'utf8');
  } catch (e) {
    console.error('Error writing feedback_log.json:', e);
  }
}

// POST /api/feedback - Save new feedback
router.post('/feedback', (req, res) => {
  const { element, comment, category, path: currentPath, viewport } = req.body;
  const list = getFeedbackList();

  const newEntry = {
    id: 'fb_' + Date.now() + '_' + Math.random().toString(36).substr(2, 5),
    timestamp: new Date().toISOString(),
    comment: comment || '(No comment provided)',
    category: category || 'General',
    element: element || {},
    currentPath: currentPath || '/',
    viewport: viewport || {}
  };

  list.unshift(newEntry);
  saveFeedbackList(list);

  console.log(`\n📢 [DIRECT BROWSER FEEDBACK RECEIVED]`);
  console.log(`💬 Comment: "${newEntry.comment}" [${newEntry.category}]`);
  console.log(`🎯 Target: <${newEntry.element.tagName?.toLowerCase()} id="${newEntry.element.id || ''}" class="${newEntry.element.className || ''}">`);
  console.log(`📍 Selector: ${newEntry.element.cssSelector}`);
  console.log(`📝 Text Snippet: "${(newEntry.element.text || '').slice(0, 100)}"`);
  console.log(`---------------------------------------------------\n`);

  res.json({ success: true, feedback: newEntry });
});

// GET /api/feedback - List all feedback
router.get('/feedback', (req, res) => {
  const list = getFeedbackList();
  res.json({ success: true, count: list.length, feedbacks: list });
});

// DELETE /api/feedback - Clear all feedback
router.delete('/feedback', (req, res) => {
  saveFeedbackList([]);
  res.json({ success: true, message: 'All feedback cleared' });
});

module.exports = router;

File: server/handlers/feedback.go
package handlers

import (
	"crypto/rand"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"net/http"
	"os"
	"path/filepath"
	"sync"
	"time"
)

var (
	feedbackMutex sync.Mutex
	feedbackFile  = "feedback_log.json"
)

type FeedbackEntry struct {
	ID          string      `json:"id"`
	Timestamp   string      `json:"timestamp"`
	Comment     string      `json:"comment"`
	Category    string      `json:"category"`
	Element     interface{} `json:"element"`
	CurrentPath string      `json:"currentPath"`
	Viewport    interface{} `json:"viewport"`
}

func getFeedbackList() []FeedbackEntry {
	feedbackMutex.Lock()
	defer feedbackMutex.Unlock()

	if _, err := os.Stat(feedbackFile); os.IsNotExist(err) {
		return []FeedbackEntry{}
	}

	data, err := os.ReadFile(feedbackFile)
	if err != nil {
		return []FeedbackEntry{}
	}

	var list []FeedbackEntry
	if err := json.Unmarshal(data, &list); err != nil {
		return []FeedbackEntry{}
	}

	return list
}

func saveFeedbackList(list []FeedbackEntry) error {
	feedbackMutex.Lock()
	defer feedbackMutex.Unlock()

	data, err := json.MarshalIndent(list, "", "  ")
	if err != nil {
		return err
	}

	dir := filepath.Dir(feedbackFile)
	if dir != "" && dir != "." {
		os.MkdirAll(dir, 0755)
	}

	return os.WriteFile(feedbackFile, data, 0644)
}

func SaveFeedbackHandler(w http.ResponseWriter, r *http.Request) {
	w.Header().Set("Content-Type", "application/json")

	var payload struct {
		Comment     string      `json:"comment"`
		Category    string      `json:"category"`
		Element     interface{} `json:"element"`
		CurrentPath string      `json:"path"`
		Viewport    interface{} `json:"viewport"`
	}

	json.NewDecoder(r.Body).Decode(&payload)

	comment := payload.Comment
	if comment == "" {
		comment = "(No comment provided)"
	}

	category := payload.Category
	if category == "" {
		category = "General"
	}

	currPath := payload.CurrentPath
	if currPath == "" {
		currPath = "/"
	}

	randBytes := make([]byte, 3)
	rand.Read(randBytes)
	id := fmt.Sprintf("fb_%d_%s", time.Now().UnixMilli(), hex.EncodeToString(randBytes))

	entry := FeedbackEntry{
		ID:          id,
		Timestamp:   time.Now().UTC().Format(time.RFC3339),
		Comment:     comment,
		Category:    category,
		Element:     payload.Element,
		CurrentPath: currPath,
		Viewport:    payload.Viewport,
	}

	list := getFeedbackList()
	newList := append([]FeedbackEntry{entry}, list...)
	saveFeedbackList(newList)

	fmt.Printf("\n📢 [DIRECT BROWSER FEEDBACK RECEIVED]\n")
	fmt.Printf("💬 Comment: \"%s\" [%s]\n", entry.Comment, entry.Category)
	fmt.Printf("---------------------------------------------------\n\n")

	json.NewEncoder(w).Encode(map[string]interface{}{
		"success":  true,
		"feedback": entry,
	})
}

func ListFeedbackHandler(w http.ResponseWriter, r *http.Request) {
	w.Header().Set("Content-Type", "application/json")
	list := getFeedbackList()

	json.NewEncoder(w).Encode(map[string]interface{}{
		"success":   true,
		"count":     len(list),
		"feedbacks": list,
	})
}

func ClearFeedbackHandler(w http.ResponseWriter, r *http.Request) {
	w.Header().Set("Content-Type", "application/json")
	saveFeedbackList([]FeedbackEntry{})

	json.NewEncoder(w).Encode(map[string]interface{}{
		"success": true,
		"message": "All feedback cleared",
	})
}
